### PR TITLE
[inductor] Support reduce over empty dimension

### PR DESCRIPTION
**Pull request by desertfire**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #88278

Summary: for https://github.com/pytorch/torchdynamo/issues/1799.
sum/all/any work on empty dimension.

cc @mlazos @soumith @voznesenskym @yanboliang @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @chunyuan-w @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx

### CODE REVIEW

File: torch/_inductor/ir.py [hunk 1]
```diff
@@ -668,6 +668,15 @@ def value_fn(index, rindex):
 
             return lambda index: fn(index)[1]
         else:
+            if all(rr == 0 for rr in reduction_ranges):
+                if reduction_type == "sum":
+                    return lambda index: ops.constant(0, dst_dtype)
+                elif reduction_type == "any":
+                    return lambda index: ops.constant(False, dst_dtype)
+                elif reduction_type == "all":
+                    return lambda index: ops.constant(False, dst_dtype)
```
Comment (jgong5) [inline torch/_inductor/ir.py:677]: `all` on empty would be `True` here?

**Comment by ngimel** [comment]
cc @krovatkin, which PR should go in, this one or #88280?

File: torch/_inductor/ir.py [hunk 2]
```diff
@@ -668,6 +668,15 @@ def value_fn(index, rindex):
 
             return lambda index: fn(index)[1]
         else:
+            if all(rr == 0 for rr in reduction_ranges):
+                if reduction_type == "sum":
+                    return lambda index: ops.constant(0, dst_dtype)
+                elif reduction_type == "any":
+                    return lambda index: ops.constant(False, dst_dtype)
+                elif reduction_type == "all":
```
Comment (ngimel) [inline torch/_inductor/ir.py:676]: there's no reduction type `all`, it's decomposed via `any` and `logical_not`

**Comment by desertfire** [comment]
Close to let Nick continuing the work.

**Pull request closed without merging**